### ISSUE TITLE
FR: Allow prioritizing MagicDNS after local DNS settings

### ISSUE THREAD

**Issue by torarnv** [tailscale/tailscale#3506]
### What are you trying to do?

I enabled MagicDNS, and was happy to see I could `ping officemachine`, but surprised that `ping home-nas` would also resolve to the TS IP. If I'm on my home network I'd rather not got through TS when interacting with machines that are available directly on that network, to reduce CPU use and increase throughput (iperf3 via TS 330 Mbits/sec, direct 440 Mbits/sec e.g.).

### How should we solve this?

It would be nice if the DNS preferences of TS would allow reordering MagicDNS after "local DNS settings". That would presumably change the search order from `torarnv.gmail.com.beta.tailscale.net mydomain` to `mydomain torarnv.gmail.com.beta.tailscale.net` and any query for `home-nas` would resolve to the local IP while `officemachine` would still end up going via TS.

I tried working around this by disabling MagicDNS, and then adding a manual split DNS entry for the TS MagicDNS domain, but the UI didn't allow me to do that:

![image](https://user-images.githubusercontent.com/7536/144756627-b4dbea3b-45b6-4463-a09c-6203c14fc43c.png)

I also tried working around it by disabling "Use Tailscale DNS settings", and then adding `torarnv.gmail.com.beta.tailscale.net` as a search domain manually in the macOS network preferences, but that didn't work. It seems the setting also shuts down the DNS server at 100.100.100.100.

### What is the impact of not solving this?

The only way I've found to work around this is to leave "Use Tailscale DNS settings" off at all times unless I'm actively trying to reach a non-local machine, which is not very pleasant.

### Anything else?

macOS 12. TS 1.18.0
```
❯ networksetup -listnetworkserviceorder
An asterisk (*) denotes that a network service is disabled.
(1) Wi-Fi
(Hardware Port: Wi-Fi, Device: en0)

(2) Thunderbolt Bridge
(Hardware Port: Thunderbolt Bridge, Device: bridge0)

(3) Tailscale Tunnel
(Hardware Port: io.tailscale.ipn.macos, Device: )
```

**Comment by torarnv** [comment]
Possible workaround: Disable MagicDNS, but leave "Use Tailscale DNS settings" enabled. This seems to still run the DNS server locally, but without messing with any DNS resolving.

Then set up manually:
```
❯ cat /etc/resolver/yourdomain.beta.tailscale.net
domainyourdomain.beta.tailscale.net
search yourdomain.beta.tailscale.net
nameserver 100.100.100.100
```

Doesn't seem to work with `nslookup`, but `ping` and other services seem to pick it up.

**Comment by torarnv** [comment]
Considering that the MagicDNS [guide](https://tailscale.com/kb/1081/magicdns/) says that:

> In the future, we will have robust enough DNS configuration and resolution logic that disabling MagicDNS separately will never be necessary. At this point, the toggle will disappear.

I hope this issue can have some kind of resolution before that point, as currently disabling MagicDNS is required to avoid directing all traffic though Trailscale for nodes in the local network.

Enabling MagicDNS puts the Tailscale domain first in the search list, and there is no way to re-order this though the UI, as you can with manually configured split-DNS.

Adding a manual split-DNS entry pointing to `100.100.100.100` results in "Invalid resolver. Only IPv4 and IPv6 values are allowed.".

The workaround is adding a DNS entry via a resolver file:

```
sudo mkdir -p /etc/resolver && sudo tee -a /etc/resolver/yourdomain.ts.net > /dev/null <<EOT
domain yourdomain.ts.net
search yourdomain.ts.net
nameserver 100.100.100.100
EOT
```

But this workaround breaks down if "Override local DNS" is checked for the global nameserver, even if an explicit split-DNS entry for the local network is added before the global nameserver in the Tailscale admin UI. Apparently redirecting any query for `foo` to `100.100.100.100` will prioritize the Tailscale domain, even if MagicDNS is disabled, and even if there are explicit split-DNS entries that would be a better match.

**Comment by torarnv** [comment]
> But this workaround breaks down if "Override local DNS" is checked for the global nameserver, even if an explicit split-DNS entry for the local network is added before the global nameserver in the Tailscale admin UI. Apparently redirecting any query for foo to 100.100.100.100 will prioritize the Tailscale domain, even if MagicDNS is disabled, and even if there are explicit split-DNS entries that would be a better match.

This turned out to be because enabling "Override local DNS" re-orders the DNS entries by setting an explicit order, e.g:

```
DNS configuration

resolver #1
  search domain[0] : foo.bar
  search domain[1] : baz.biz
  nameserver[0] : 100.100.100.100
  if_index : 19 (utun5)
  flags    : Supplemental, Request A records, Request AAAA records
  reach    : 0x00000003 (Reachable,Transient Connection)
  order    : 100200

resolver #2
  nameserver[0] : 1.2.3.4
  if_index : 6 (en0)
  flags    : Request A records
  reach    : 0x00020002 (Reachable,Directly Reachable Address)
  order    : 200000

resolver #3
  domain   : baz.biz.
  nameserver[0] : 100.100.100.100
  if_index : 19 (utun5)
  flags    : Supplemental, Request A records, Request AAAA records
  reach    : 0x00000003 (Reachable,Transient Connection)
  order    : 100201

resolver #4
  domain   : foo.bar.
  nameserver[0] : 100.100.100.100
  if_index : 19 (utun5)
  flags    : Supplemental, Request A records, Request AAAA records
  reach    : 0x00000003 (Reachable,Transient Connection)
  order    : 100202
```

This can be worked around by adding an explicit order to `/etc/resolver/yourdomain.ts.net` that's after all the other resolvers, i.e.:

```
sudo mkdir -p /etc/resolver && sudo tee -a /etc/resolver/yourdomain.ts.net > /dev/null <<EOT
domain yourdomain.ts.net
search yourdomain.ts.net
nameserver 100.100.100.100
search_order 1000000
EOT
```

**Comment by ryanwwest** [comment]
+1. I'd like to always use the same Tailscale hostname as well, whether or not I'm on the local IP, but iperf tunnel traffic throughput is much slower vs. connecting to the local IP directly. I'd prefer to not have to make any DNS changes at all and just have the hostname resolve to the local IP when the user prefers it to utilize the maximum bandwidth and minimum latency.

**Comment by Cubox** [comment]
Would love an option to just say "don't set the domain search value in my DNS records but keep everything else"

**Comment by DentonGentry** [comment]
> Would love an option to just say "don't set the domain search value in my DNS records but keep everything else"

https://tailscale.com/kb/1054/dns/#search-domains provides a way to control whether a search domain is added.

**Comment by Cubox** [comment]
> > Would love an option to just say "don't set the domain search value in my DNS records but keep everything else"
> 
> https://tailscale.com/kb/1054/dns/#search-domains provides a way to control whether a search domain is added.

What we need is to disable this:
    When MagicDNS is enabled, it is always the first domain in the Search Domains list. This is not configurable by a user.

**Comment by DentonGentry** [comment]
100.100.100.100 is a tiny DNS server running within the Tailscale process on the local device. It directly answers queries for the MagicDNS FQDN(s), and forwards other queries off to other DNS servers. Importantly, if deliberately sent a query it will always send a response. Even if Tailscale DNS is disabled such that nothing should be sending queries to it, deliberately sending a query to 100.100.100.100 will get a response.

One can get complete control of the DNS configuration by disabling Tailscale DNS, either in https://login.tailscale.com/admin/dns by disabling all features or on the local device by turning off "Accept Tailscale DNS" in the Preferences menu (or using `--accept-dns=false` from the CLI).

One can then add your own DNS config to send tails0123456.ts.net to 100.100.100.100, and it will get a response.

**Comment by Cubox** [comment]
> 100.100.100.100 is a tiny DNS server running within the Tailscale process on the local device. It directly answers queries for the MagicDNS FQDN(s), and forwards other queries off to other DNS servers. Importantly, if deliberately sent a query it will always send a response. Even if Tailscale DNS is disabled such that nothing should be sending queries to it, deliberately sending a query to 100.100.100.100 will get a response.
> 
> One can get complete control of the DNS configuration by disabling Tailscale DNS, either in https://login.tailscale.com/admin/dns by disabling all features or on the local device by turning off "Accept Tailscale DNS" in the Preferences menu (or using `--accept-dns=false` from the CLI).
> 
> One can then add your own DNS config to send tails0123456.ts.net to 100.100.100.100, and it will get a response.

Hi,

Everything you are saying is true, however the issue lies elsewhere.

What is impacting us is the fact that the search domain is set and cannot be removed.
Let's say I have on my network a node named "Rhea", with MagicDNS. If I try to reach Rhea, it will go through TS. However, I'm on my local network, and I don't want to use TS.
If I want to go through TS, I want to be able to still use rhea.name.ts.net to go through TS.

Basically, all that is required is the option to disable the "search name.ts.net" from being added to resolv.conf (or equivalent on other OSes), while still setting nameserver 100.100.100.100

Also, the current behaviour, this breaks the mail daemon on my server. It tries to send an email to root@onyx, which due to the search domain means it tries onyx.domain.ts.net instead of just onyx, which is localhost.

Let me know if I'm unclear, I'd love for this fix to be implemented, as this is annoying and has me use raw IPs everywhere.

**Comment by DentonGentry** [comment]
> Basically, all that is required is the option to disable the "search name.ts.net" from being added

I understand the request.

I am saying that we are unlikely to implement a configuration option which would be a footgun to most users, those who are less sophisticated about how DNS works and do not have detailed requirements about how their search domains should work. If they remove the MagicDNS search domain, they then contact support. We aim for simple things to be easy and more complex things to be _possible_. We avoid providing a plethora of configuration options.

To have complete control over search domains and ordering:
- 100.100.100.100 will answer any query which is sent to it, whether Magic DNS is on or off.
- you can configure /etc/resolv.conf however you'd prefer.

**Comment by Cubox** [comment]
Could a command line only flag to enable this behaviour be added? IIRC, there are options that are command line only. This would solve the unaware users contacting support issue

**Comment by torarnv** [comment]
@DentonGentry 

You don't have add the option to _remove_  the `name.ts.net` search domain, just allow it to be _re-ordered_ (to the end if needed), so that local search domains can be prioritized?

**Comment by yanii** [comment]
My workaround for this that doesn't require disabling magicdns, assuming you have a local search domain you want to prioritize over Tailscale on some servers, e.g. ``local.domain``, is to change the domain name search order on the servers. Much of this will depend on your Linux distribution/version, but below is for Ubuntu 22.04.

Edit ``/etc/systemd/resolved.conf``, uncomment and manually set the ``Domains=`` line to include your local domain:
```
...
#FallbackDNS=
Domains=local.domain
#DNSSEC=no
...
```

and then regenerate /etc/resolv.conf:

```
sudo systemctl restart systemd-resolved
```

Finally, check your domain search order is now correct in ``/etc/resolv.conf``:

```
search local.domain something.ts.net 
```

**Comment by torarnv** [comment]
@DentonGentry @bradfitz Sorry to bring this up again, but I'm hoping there will be a solution to this down the road.

I'd also like to stress that the [MagicDNS documentation](https://tailscale.com/kb/1081/magicdns) still notes that:

> _In the future, we will have robust enough DNS configuration and resolution logic that disabling MagicDNS separately will never be necessary. At this point, the toggle will disappear._

As far as I can tell disabling MagicDNS is the only solution to this issue for now, so please don't remove that ability 🙏

For example, the very interesting solution in https://willnorris.com/2023/tailscale-custom-domain/ relies on this as well I think, at least to support `foo` → `foo.ipn.willnorris.net` (Correct me if I'm wrong @willnorris).

**Comment by anacrolix** [comment]
+1 — this is still a problem as of macOS Sequoia with Tailscale 1.78+.

In my case, enabling MagicDNS causes Tailscale to register a Supplemental resolver on the `utun` interface with `order: 102200` that claims the search domains `local`, `lan`, and the tailnet domain. The local gateway's resolver (`192.168.0.1` on `en0`) gets pushed into the scoped-only section of `scutil --dns`, meaning it's never consulted for general lookups.

The practical impact: local hostnames like `myhost.localdomain` or `myhost.lan` (served by the LAN's gateway) stop resolving because Tailscale's `100.100.100.100` doesn't know about them. Disabling Tailscale immediately restores local resolution.

The current workaround is either:
1. `/etc/resolver/` overrides on each macOS machine (doesn't scale, doesn't survive OS updates)
2. Split DNS in the admin console (works cross-device but requires knowing all local domains upfront)

Neither is a substitute for having the option to set MagicDNS as a lower-priority resolver that only handles `*.ts.net` queries, falling back to the local resolver for everything else. This would make Tailscale a good network citizen on LANs with existing DNS infrastructure.
